src/NodeServices/PythonServices.cs

```diff
@@ -111,12 +111,52 @@ public sealed class PythonEngineManager
         public static PythonEngineManager Instance { get { return lazy.Value; } }
         #endregion
 
-        //TODO see DYN-6550 when hiding/replacing this obsolete field.
+        /// <summary>
+        /// A readonly collection of all the loaded Python engines
+        /// </summary>
+        public ReadOnlyCollection<PythonEngine> PythonEngines => new ReadOnlyCollection<PythonEngine>(AvailableEngines);
+
         /// <summary>
         /// An observable collection of all the loaded Python engines
         /// </summary>
-        [Obsolete("AvailableEngines field will be replaced in a future Dynamo release.")]
-        public ObservableCollection<PythonEngine> AvailableEngines;
+        internal ObservableCollection<PythonEngine> AvailableEngines;
+
+        public delegate void PythonEngineChangedHandler(PythonEngine engine);
+
+        private Action<PythonEngine> customizeEngine;
+
+        /// <summary>
+        /// Use this function to customize Python engine initialization.
+        /// This function will be called only once on all existing or future python engines (on PythonEngineAdded).
+        /// </summary>
+        public Action<PythonEngine> CustomizeEngine
+        {
+            set
+            {
+                customizeEngine = value;
+                if (customizeEngine != null)
+                {
+                    try
+                    {
+                        foreach (var engine in AvailableEngines)
+                        {
+                            customizeEngine(engine);
+                        }
+                    }
+                    catch { }
+                }
+            }
+        }
+
+        /// <summary>
+        /// Event that is triggered for every new engine that is added.
+        /// </summary>
+        public static PythonEngineChangedHandler PythonEngineAdded;
+
+        /// <summary>
+        /// Event that is triggered for every engine that is removed.
+        /// </summary>
+        public static PythonEngineChangedHandler PythonEngineRemoved;
```

Comment:
how do you imagine removal will work? Just remove the instance from the list of engines or actually try to unload it?

Reply:
With ALCs, I would imagine completely unloading it